README.md
# FMC_Unsupported_Algorithms - Update 1.0.6!
[![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/RenanHingel/fmc_unsupported_algorithms)[![Run in Cisco Cloud IDE](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-runable-icon.svg)](https://developer.cisco.com/devenv/?id=devenv-vscode-base&GITHUB_SOURCE_REPO=https://github.com/RenanHingel/fmc_unsupported_algorithms)

## Recent changes:

Oct 16 - Version 1.0.6:
+ Fixed faulty IKE and IPsec lists

Oct 16 - Version 1.0.5:
+ Reworked IKEv1 IKE Policy review
+ Encryption Algorithms review
+ Removed debug logging, instead, logging function is verbose aware now - future use

[View more on the changelog file](changelog.txt)

## The motivator:
To elevate the security standards, Cisco updated the acceptable algorithms that may be used to build VPN tunnels starting from firmware 6.7 and superior versions.

    You may not be able to upgrade an FMC, or deploy post-upgrade with FDM, if you use any of the following FTD features:

    Diffie-Hellman groups: 2, 5, and 24.
    Group 5 continues to be supported in FMC deployments for IKEv1, but we recommend you change to a stronger option.
    Encryption algorithms for users who satisfy export controls for strong encryption: DES, 3DES, AES-GMAC, AES-GMAC-192, AES-GMAC-256. DES continues to be supported (and is the only option) for users who do not satisfy export controls.
    The NULL "encryption algorithm" (authentication without encryption, for testing purposes) continues to be supported in FMC deployments for both IKEv1 and IKEv2 IPsec proposals. However, it is no longer supported in IKEv2 policies.
    Hash algorithms: MD5.

    If you are still using these features in IKE proposals or IPsec policies, change and verify your VPN configuration before you upgrade. 
    
Source: [https://www.cisco.com/c/en/us/support/docs/security-vpn/ipsec-negotiation-ike-protocols/216276-configure-route-based-site-to-site-vpn-t.html](https://www.cisco.com/c/en/us/td/docs/security/firepower/670/relnotes/firepower-release-notes-670/m_features_functionality.html?bookSearch=true)

If you have just a few VPNs configured, determining which VPNs must be modified is an easy task. 
However, if you need to generate a structured report and contact dozens (and sometimes hundreds) of VPN peers to inform them about the need to update their end, this task can consume hundreds of hours and examining the configurations manually is prone to human error.

The managed devices still running with these deprecated settings are unable to be upgraded past firmware 6.6.X, this poses a stability and security threat to your network environment.

## A proposed solution:
Use the Cisco Secure Firewall Management Center (FMC) API to leverage the information about VPNs, automatically parse the unsecure configurations and create a log file with completed run details and a clean CSV file containing just the important information.

![Screenshot](img_script.png)

## Understanding the script logic (UML):

          +-----------------------------------------------------+
          |              Start Program                          |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Generate FMC token                                 |
          |  /api/fmc_platform/v1/auth/generatetoken            |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Get a full list of all IKEv1 IPSEC Proposals       |
          |  /api/fmc_config/v1/domain/{domainUUID}/object/     |
          |  ikev1ipsecproposals?offset=0&limit=4&expanded=true |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Get a full list of all IKEv2 IPSEC Proposals       |
          |  /api/fmc_config/v1/domain/{domainUUID}/object/     |
          |  ikev2ipsecproposals?offset=0&limit=10&expanded=true|
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Get a full list of all IKEv1 IKE Policies          |
          |  /api/fmc_config/v1/domain/{domainUUID}/object/     |
          |  ikev1policies?offset=0&limit=10&expanded=true      |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Get a full list of all IKEv2 IKE Policies          |
          |  /api/fmc_config/v1/domain/{domainUUID}/object/     |
          |  ikev2policies?offset=0&limit=10&expanded=true      |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Loop (For Each IKEv1 and IKEv2 IKE Policy)         |
          |                                                     |
          |    +-----------------------------------------------------------+
          |    |  Launch GET Request to Obtain IKE Settings                |
          |    |  /api/fmc_config/v1/domain/{domainUUID}/policy/           |
          |    |  ftds2svpns/{objectId}/ikesettings/{ike_settings_id}      |
          |    +-----------------------------------------------------------+
          |                                                     |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  Loop (For Each VPN)                                |
          |                                                     |
          |    +-----------------------------------------------------------+
          |    |  Launch GET Request to Obtain VPN Details                 |
          |    |  /api/fmc_config/v1/domain/{domainUUID}/policy/           |
          |    |  ftds2svpns/{vpn_id}/ikesettings/{ike_settings_id}        |
          |    +-----------------------------------------------------------+
          |                                                     |
          +-----------------------------------------------------+
                V
          +-----------------------------------------------------+
          |  End Program                                        |
          +-----------------------------------------------------+


## Solution Components
* Cisco FMC API
* Python

## Installation:

### Console
#### Clone the repo
```console
git clone https://github.com/RenanHingel/fmc_unsupported_algorithms.git
```

### Python

#### Set up a Python venv
First make sure that you have Python 3 installed on your machine. We will then be using venv to create
an isolated environment with only the necessary packages.

##### Install virtualenv via pip
```
pip install virtualenv
```

##### Create a new venv
```
+++ Change to your project folder
cd fmc_unsupported_algorithms

+++ Create the venv
virtualenv venv

+++ Activate your venv
source venv/bin/activate
```

#### Install dependencies
```zsh
pip3 install -r requirements.txt
```

#### Run the script
```
python3 get_vpn_requirements.py
```

## Contacts
* Renan Hingel (renanhingel@gmail.com)


changelog.txt
## Changelog:

Version 1.0.6:
+ Fixed faulty IKE and IPsec lists

Version 1.0.5:
+ Reworked IKEv1 IKE Policy review
+ Encryption Algorithms review
+ Removed debug logging, instead, logging function is verbose aware now - future use

Version 1.0.4:
+ Fixed IKEv1 policy parsing
+ Added one second delay to prevent rate limiting issues

Version 1.0.3
+ Added limit=1000 to handle pagination
+ Added debug logging
+ Added IKEv1 support

Version 1.0.2
+ Added a try/except layer to expect IKEv2 IPsec Proposal: NONE

Version 1.0.0 and 1.0.1
+ Complete chaos, don't rollback to these :-)


get_vpn_info.py
########################################################
# FMC_Unsupported_Algorithms
#
# Changelog for current version 1.0.6:
# + Fixed faulty IKE and IPsec lists
#
# For full version control, please refer to https://github.com/RenanHingel/fmc_unsupported_algorithms
########################################################

import time
import requests
import urllib3
import getpass
from requests.auth import HTTPBasicAuth
import datetime
import csv
urllib3.disable_warnings()

version = "1.0.6"

address = input("Enter the FMC IP address: ")
username = input("Enter your username: ")
password = getpass.getpass(prompt="Enter your password: ")
base_url = "https://" + address
verbose = True

# Declaring variables
review_ike_policies = {}
vpn_data_list = []
todays_date = datetime.datetime.now()
logfile = f"output_DATE{todays_date.day}{todays_date.month}{todays_date.year}_TIME{todays_date.hour}{todays_date.minute}.txt"
log = open(logfile,"a")


def log_and_print(input_string, verbose):
    if verbose == True:
        print(input_string)
    log.write(input_string + '\n')

def api_call_get(api_url):
    try:
        response = requests.get(api_url, headers=headers, verify=False)  # Since it's an example, we'll disable SSL verification.
        response.raise_for_status()  # Check if the request was successful (status code 200-299).
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error during API request: {e}")
        return None

def review_ike_policy(policy, review, version):
    if version == 1:
        encryption_search = "encryption"
        dh_search = "diffieHellmanGroup"
    if version == 2:
        encryption_search = "encryptionAlgorithms"
        dh_search = "diffieHellmanGroups"

    invalid_algorithms = ['3DES', 'AES-GMAC', 'AES-GMAC-192', 'AES-GMAC-256', 'DES', 'NULL']
    required_groups = [2, 5, 24]

    invalid_algorithms_found = [algorithm for algorithm in invalid_algorithms if algorithm in policy[encryption_search]]
    if invalid_algorithms_found:
        review[policy['name']] = {'Encryption': invalid_algorithms_found}

    if isinstance(policy[dh_search], int):
        policy[dh_search] = [policy[dh_search]]

    dh_groups_found = [group for group in policy[dh_search] if group in required_groups]
    if dh_groups_found:
        if policy['name'] in review:
            review[policy['name']]['DH group'] = dh_groups_found
        else:
            review[policy['name']] = {'DH group': dh_groups_found}

def review_ipsec_proposals(proposal_data, version):
    review_result = {}
    invalid_algorithms = ['3DES', 'AES-GMAC', 'AES-GMAC-192', 'AES-GMAC-256', 'DES', 'NULL']

    for item in proposal_data['items']:
        proposal_name = item['name']
        invalid_algorithms_found = []

        if version == 1:
            esp_encryption = item["espEncryption"]
            if esp_encryption in invalid_algorithms:
                invalid_algorithms_found.append(esp_encryption)

        if version == 2:
            encryption_algorithms = item["encryptionAlgorithms"]
            for algorithm in encryption_algorithms:
                if algorithm in invalid_algorithms:
                    invalid_algorithms_found.append(algorithm)

        if invalid_algorithms_found:
            review_result[proposal_name] = {'Encryption': invalid_algorithms_found}
    
    return review_result


# Script header
log_and_print(f"==========================================================", True)
log_and_print(f"FMC Unsupported Algorithms - Version {version}", True)
log_and_print(f"This script aims to leverage Cisco FMC API to look for:", verbose)
log_and_print(f"- IKEv1/v2 Policy: Diffie-Hellman groups 2, 5 and 24.", verbose)
log_and_print(f"- IKEv1/v2 Policy & IPsec Proposal containing encryption algorithms '3DES', 'AES-GMAC', 'AES-GMAC-192', 'AES-GMAC-256', 'DES' and 'NULL'.", verbose)
log_and_print(f"Source: https://www.cisco.com/c/en/us/td/docs/security/secure-firewall/management-center/device-config/720/management-center-device-config-72/vpn-s2s.html", verbose)
log_and_print(f"----------------------------------------------------------", True)
log_and_print(f"Connecting to FMC {base_url}", True)

# Start of FMC access token token generation
token_uri = "/api/fmc_platform/v1/auth/generatetoken"
response = requests.request("POST", base_url + token_uri, verify=False, auth=HTTPBasicAuth(username, password))
accesstoken = response.headers["X-auth-access-token"]
refreshtoken = response.headers["X-auth-refresh-token"]
DOMAIN_UUID = response.headers["DOMAIN_UUID"]
log_and_print(f"Connection status code: {response.status_code}", True)

# Define headers for all subsequent API calls
headers = { 'Content-Type': 'application/json', 'x-auth-access-token': accesstoken }

# List of FMC API endpoints used by this script
ikev1_ipsecproposals_uri = "/api/fmc_config/v1/domain/" + DOMAIN_UUID + "/object/ikev1ipsecproposals?offset=0&limit=4&expanded=true"
ikev2_ipsecproposals_uri = "/api/fmc_config/v1/domain/" + DOMAIN_UUID + "/object/ikev2ipsecproposals?offset=0&limit=10&expanded=true"
ikev1_policies_uri = "/api/fmc_config/v1/domain/" + DOMAIN_UUID + "/object/ikev1policies?offset=0&limit=10&expanded=true"
ikev2_policies_uri = "/api/fmc_config/v1/domain/" + DOMAIN_UUID + "/object/ikev2policies?offset=0&limit=10&expanded=true"

# Get IPSEC proposals data
ikev1ipsecproposals_data = api_call_get(base_url + ikev1_ipsecproposals_uri)
ikev2ipsecproposals_data = api_call_get(base_url + ikev2_ipsecproposals_uri)

# Get IKE policies data
ikev1policies_data = api_call_get(base_url + ikev1_policies_uri)
ikev2policies_data = api_call_get(base_url + ikev2_policies_uri)

# Iterate through the items - IKEv1 and IKEv2 IKE Policy
for item in ikev1policies_data['items']:
    review_ike_policy(item, review_ike_policies, 1)
for item in ikev2policies_data['items']:
    review_ike_policy(item, review_ike_policies, 2)

# Iterate through the items - IKEv1 and IKEv2 IPSEC Proposals
ikev1_ipsec_review_result = review_ipsec_proposals(ikev1ipsecproposals_data, 1)
ikev2_ipsec_review_result = review_ipsec_proposals(ikev2ipsecproposals_data, 2)

# Now that we know which IKE policies and IPsec proposals need to be fixed, we will pull the complete VPN list for this FMC
all_vpn_uri = "/api/fmc_config/v1/domain/e276abec-e0f2-11e3-8169-6d9ed49b625f/policy/ftds2svpns?limit=1000&expanded=true"
all_vpn_data = api_call_get(base_url + all_vpn_uri)

# Then, for each VPN configured, we will find it's IKE and IPsec information, then match it against the audit we did previously
for item in all_vpn_data["items"]:
        ipsec_fix_list = []
        ike_fix_list= []
        current_vpn_ipsec_list = []
        current_vpn_ike_policy_list = []

        vpn_name = item["name"]
        vpn_id = item["id"]

        isikev1 = item["ikeV1Enabled"]
        isikev2 = item["ikeV2Enabled"]

        ike_settings_id = item['ikeSettings']['id']
        ike_settings_uri = base_url + "/api/fmc_config/v1/domain/" + DOMAIN_UUID + f"/policy/ftds2svpns/{vpn_id}/ikesettings/{ike_settings_id}"
        ike_settings_data = api_call_get(ike_settings_uri)

        if isikev1 == True:
            for proposal in item["ipsecSettings"]["ikeV1IpsecProposal"]:
                proposal_name = proposal["name"]
                current_vpn_ipsec_list.append(proposal_name)
            ikev1_policies = [policy['name'] for policy in ike_settings_data['ikeV1Settings']['policies']]
            current_vpn_ike_policy_list.extend(ikev1_policies)
        
        if isikev2 == True:
            for proposal in item["ipsecSettings"]["ikeV2IpsecProposal"]:
                proposal_name = proposal["name"]
                current_vpn_ipsec_list.append(proposal_name)
            ikev2_policies = [policy['name'] for policy in ike_settings_data['ikeV2Settings']['policies']]
            current_vpn_ike_policy_list.extend(ikev2_policies)
       

        log_and_print("==========================================================", True)
        log_and_print(f"|--- VPN Name: {vpn_name}", True)
        log_and_print(f"|    |--- IPsec Proposals: {', '.join(current_vpn_ipsec_list)}", True)
        log_and_print(f"|    |--- IKE Proposals: {', '.join(current_vpn_ike_policy_list)}", True)

        for item in current_vpn_ipsec_list:
            if item in ikev1_ipsec_review_result:
                ikev1_encryption_info = ', '.join(ikev1_ipsec_review_result[item]['Encryption'])
                ike_review_info = f"Review needed - IKEv1 IPsec proposal {item} - REMOVED encryption: {ikev1_encryption_info}"
                ipsec_fix_list.append(ike_review_info)
                log_and_print(ike_review_info, True)
            if item in ikev2_ipsec_review_result:
                ikev2_encryption_info = ', '.join(ikev2_ipsec_review_result[item]['Encryption'])
                ike_review_info = f"Review needed - IKEv2 IPsec proposal {item} - REMOVED encryption: {ikev2_encryption_info}"
                ipsec_fix_list.append(ike_review_info)
                log_and_print(ike_review_info, True)

        for item in current_vpn_ike_policy_list:
            if item in review_ike_policies:
                policy_info = review_ike_policies[item]
                encryption_info = ', '.join(policy_info.get('Encryption', []))
                dh_group_info = ', '.join(str(group) for group in policy_info.get('DH group', []))

                ike_review_info = f"Review needed for {item}"
                if encryption_info:
                    ike_review_info += f" - REMOVED encryption: {encryption_info}"
                if dh_group_info:
                    ike_review_info += f" - REMOVED DH Groups: {dh_group_info}"
                ike_fix_list.append(ike_review_info)
                log_and_print(ike_review_info, True)
        vpn_data_list.append([vpn_name,', '.join(current_vpn_ike_policy_list),', '.join(ike_fix_list),', '.join(current_vpn_ipsec_list),', '.join(ipsec_fix_list)])
        time.sleep(1)
print("==========================================================")

csv_filename = f"vpn_report_DATE{todays_date.day}{todays_date.month}{todays_date.year}_TIME{todays_date.hour}_{todays_date.minute}.csv"
with open(csv_filename, 'w', newline='') as csv_file:
    csv_writer = csv.writer(csv_file)
    csv_writer.writerow(['vpn_name', 'IKE_policies','IKE_fix_list', 'IPSEC_policies','IPSEC_fix_list'])
    csv_writer.writerows(vpn_data_list)

log_and_print(f"Collected information written to: {logfile}", True)
log_and_print(f"Collected information written to: {csv_filename}", True)
log_and_print(f"==========================================================", True)
